src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java

```diff
@@ -27,12 +35,17 @@
 @Component
 public class MenuSessionFactory {
 
+    private static final String NEXT_SCREEN = "NEXT_SCREEN";
```

Comment:
Do you mean the whole rebuild function or the addition of the queries? I guess I don't really understand why the rebuild function is needed at all.

Reply:
It seems like 'NEXT_SCREEN" was introduced as a formal type of stack frame step (IE: It sets the command code to a new concept, but only in formplayer, so I was hoping to understand the semantics a bit better.

"Rebuild" is just intended to be a way to encode the "meaning' of the current frame steps into a central place, rather than making it based on event-driven changes. IE: If the "current command" depends on interpreting all of the stack frames together, setting the outcome via a rebuild lets the engine put all the logic for how to make the decision in one place.

Reply:
This is an artefact of the way query screens are processed in the `MenuSession`. In normal navigation we don't treat query screens as navigable steps, instead we treat them as intermediate steps and don't record any state about them in the user selections (they query data is stored separately).

When adding queries to the stack this means we have to also take a different approach than what we do for other screens.

There are other side effects of this. Currently we could not have multiple query screens that need user input at the same command level (since we associate the query data with the current command).